File: src/Heq/Storage/Correction.cs
using System.Collections.Generic;
using System.Linq;
using Heq.Model;

namespace Heq.Storage
{
    public class Correction
    {
        public List<BandDto> Bands { get; set; } = new List<BandDto>();

        public bool Enabled { get; set; } = true;

        public bool IsEmpty => Bands.Count == 0;

        public bool Applies => Enabled && Bands.Any(b => b.Enabled);

        public static Correction FromModel(EqModel m)
            => new Correction { Bands = m.Bands.Select(BandDto.From).ToList() };

        public void ApplyTo(EqModel m)
        {
            using (m.Batch())
            {
                m.Bands.Clear();
                foreach (var b in ToBands()) m.Bands.Add(b);

                m.AutoGain = false;
                m.PreampDb = 0;
                m.Touch();
            }
        }

        public IEnumerable<EqBand> ToBands()
        {
            foreach (var d in Bands)
            {
                var b = d.ToBand();
                if (b.Channel == ChannelTarget.Both) b.Channel = ChannelTarget.Left;
                yield return b;
            }
        }

        public string Summary()
        {
            int left = Bands.Count(b => b.Channel != ChannelTarget.Right);
            int right = Bands.Count - left;
            return Bands.Count == 0 ? "no correction" : $"{left} left · {right} right";
        }
    }
}

File: rust/heq-core/src/storage/correction.rs
use serde::{Deserialize, Serialize};

use super::BandDto;
use crate::model::{ChannelTarget, EqBand, EqModel};

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct Correction {
    #[serde(default)]
    pub bands: Vec<BandDto>,
    #[serde(default = "default_true")]
    pub enabled: bool,
}

fn default_true() -> bool {
    true
}

impl Default for Correction {
    fn default() -> Self {
        Correction {
            bands: Vec::new(),
            enabled: true,
        }
    }
}

impl Correction {
    pub fn from_model(m: &EqModel) -> Self {
        Correction {
            bands: m.bands().iter().map(BandDto::from_band).collect(),
            enabled: true,
        }
    }

    pub fn is_empty(&self) -> bool {
        self.bands.is_empty()
    }

    pub fn applies(&self) -> bool {
        self.enabled && self.bands.iter().any(|b| b.enabled)
    }

    pub fn apply_to(&self, m: &mut EqModel) {
        m.batch(|m| {
            m.clear();
            for b in self.to_bands() {
                m.push_band(b);
            }
            m.set_auto_gain(false);
            m.set_preamp_db(0.0);
            m.touch();
        });
    }

    pub fn to_bands(&self) -> Vec<EqBand> {
        self.bands
            .iter()
            .map(|d| {
                let mut b = d.to_band();
                if b.channel == ChannelTarget::Both {
                    b.channel = ChannelTarget::Left;
                }
                b
            })
            .collect()
    }

    pub fn summary(&self) -> String {
        if self.bands.is_empty() {
            return "no correction".to_string();
        }
        let left = self
            .bands
            .iter()
            .filter(|b| b.channel != ChannelTarget::Right)
            .count();
        format!("{} left · {} right", left, self.bands.len() - left)
    }
}
